swarm/storage/pyramid.go

```diff
@@ -73,129 +131,486 @@ func NewPyramidChunker(params *ChunkerParams) (self *PyramidChunker) {
 	self.branches = params.Branches
 	self.hashSize = int64(self.hashFunc().Size())
 	self.chunkSize = self.hashSize * self.branches
-	self.workerCount = 1
+	self.workerCount = 0
+
 	return
 }
 
-func (self *PyramidChunker) Split(data io.Reader, size int64, chunkC chan *Chunk, swg, wwg *sync.WaitGroup) (Key, error) {
+func (self *PyramidChunker) Join(key Key, chunkC chan *Chunk) LazySectionReader {
+
+	return &LazyChunkReader{
+		key:       key,
+		chunkC:    chunkC,
+		chunkSize: self.chunkSize,
+		branches:  self.branches,
+		hashSize:  self.hashSize,
+	}
+}
+func (self *PyramidChunker) Split(data io.Reader, size int64, chunkC chan *Chunk, storageWG, processorWG *sync.WaitGroup) (Key, error) {
+
+	jobC := make(chan *chunkJob, 2*ChunkProcessors)
+	wg := &sync.WaitGroup{}
+	errC := make(chan error)
+	quitC := make(chan bool)
+	rootKey := make([]byte, self.hashSize)
+	chunkLevel := make([][]*TreeEntry, self.branches)
 
-	chunks := (size + self.chunkSize - 1) / self.chunkSize
-	depth := int(math.Ceil(math.Log(float64(chunks))/math.Log(float64(self.branches)))) + 1
+	wg.Add(1)
+	go self.prepareChunks(false, chunkLevel, data, rootKey, quitC, wg, jobC, processorWG, chunkC, errC, storageWG)
 
-	results := Tree{
-		Chunks: chunks,
-		Levels: make([]map[int64]*Node, depth),
+	// waiting for all chunks to finish
+	wg.Wait()
+
+	// if storage waitgroup is non-nil, we wait for storage to finish too
+	if storageWG != nil {
+		storageWG.Wait()
 	}
-	for i := 0; i < depth; i++ {
-		results.Levels[i] = make(map[int64]*Node)
+	close(errC)
+	close(quitC)
+
+	select {
+	case err := <-errC:
+		if err != nil {
+			close(quitC)
+			return nil, err
+		}
+		//TODO: add a timeout
 	}
-	// Create a pool of workers to crunch through the file
-	tasks := make(chan *Task, 2*processors)
-	pend := new(sync.WaitGroup)
-	abortC := make(chan bool)
-	for i := 0; i < processors; i++ {
-		pend.Add(1)
-		go self.processor(pend, swg, tasks, chunkC, &results)
+	return rootKey, nil
+
+}
+
+func (self *PyramidChunker) Append(key Key, data io.Reader, chunkC chan *Chunk, storageWG, processorWG *sync.WaitGroup) (Key, error) {
+
+	quitC := make(chan bool)
+	rootKey := make([]byte, self.hashSize)
+	chunkLevel := make([][]*TreeEntry, self.branches)
+
+	// Load the right most unfinished tree chunks in every level
+	self.loadTree(chunkLevel, key, chunkC, quitC)
+
+	jobC := make(chan *chunkJob, 2*ChunkProcessors)
+	wg := &sync.WaitGroup{}
+	errC := make(chan error)
+
+	wg.Add(1)
+	go self.prepareChunks(true, chunkLevel, data, rootKey, quitC, wg, jobC, processorWG, chunkC, errC, storageWG)
+	// waiting for all chunks to finish
+	wg.Wait()
```

Comment:
plus the same error as the one you fixed above with `close(quit)` not alwayscalled